### DISCUSSION TITLE
Getting a notification when Rocket is shutting down?

### DISCUSSION THREAD

**Discussion by darksylinc** [rwf2/Rocket#2800] [Questions] [answered]
Hi!

I need to save certain transient data to disk on shutdown. By "transient data" I mean:

 - "It would be really cool if this data is saved so it can be restored on server restart".
 - It's tolerable (but undesirable) if all data is lost.
 - It's completely fine if the last \~60 seconds of collected data are lost.

Does Rocket support such functionality? (if not, I can open a ticket).

I can see that `rocket::Config::shutdown.signals` might be close to what I'm looking for. But it appears to be low level and exposing implementation details?

Basically all I need would be among the terms of:

```rs
fn my_shutdown_routine( my_data: &rocket::State<MyData> ) {
   // Alright. Rocket is going down. Time to write everything to disk.
} 

#[launch]
fn rocket() -> _ {
   rocket::build().manage(my_data).shutdown(my_shutdown_routine);
   /* ... */
}
```

Ideally `my_shutdown_routine` should not be part of the grace period.

Did I miss this functionality in the docs?

Cheers and thanks in advance.

**Comment by wezm** [comment, marked as answer]
I think this is what you're after: https://api.rocket.rs/v0.5/rocket/fairing/struct.AdHoc#method.on_shutdown with example use here:

https://rocket.rs/guide/v0.5/fairings/#ad-hoc-fairings

**Comment by darksylinc** [comment]
Thank you!

That's exactly what I was looking for and it worked!

The final code looks like this:

```rs
rocket::build()
	.attach(rocket::fairing::AdHoc::on_shutdown("Bye!", |r| {
	    Box::pin(async move {
		let r = r.state::<MyData>().unwrap();
		println!("Rocket is on its way back!");
	    })
	}))
	.manage(my_data)
	.mount( /* ... */ )
```
